server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -571,6 +571,44 @@ public Snapshot newSnapshotFromGen(long minGeneration) throws IOException {
         }
     }
 
+    /**
+     * Reads and returns the operation from the given location if the generation it references is still available. Otherwise
+     * this method will return <code>null</code>.
+     */
+    public Operation readOperation(Location location) throws IOException {
+        BaseTranslogReader reader = null;
+        Closeable onClose = null;
+        try (ReleasableLock ignored = readLock.acquire()) {
+            ensureOpen();
+            if (location.generation < getMinFileGeneration()) {
+                return null;
+            }
+            if (current.generation == location.generation) {
+                // fsync here to ensure all buffers are written to disk
+                current.syncUpTo(location.translogLocation + location.size);
+                reader = current;
+                onClose = acquireTranslogGenFromDeletionPolicy(current.generation);
+
+            } else {
+                for (int i = readers.size() - 1; i >= 0; i--) {
+                    TranslogReader translogReader = readers.get(i);
+                    if (translogReader.generation == location.generation) {
+                        reader = translogReader;
+                        onClose = acquireTranslogGenFromDeletionPolicy(current.generation);
```

Comment:
closing looks to be an expensive call, because upon closing, the write lock is acquired twice (once in trimUnreferencedReaders, and once in closeFilesIfNoPendingRetentionLocks). I wonder if we need to optimize those two methods now.

Reply:
I will look into it